include/umf/providers/provider_coarse.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2023-2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#ifndef UMF_COARSE_PROVIDER_H
+#define UMF_COARSE_PROVIDER_H
+
+#include <stdbool.h>
+#include <umf/memory_provider.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// @brief Coarse Memory Provider settings struct.
+typedef struct coarse_memory_provider_params_t {
+    /// Handle to the upstream memory provider.
+    /// It has to be NULL if init_buffer is set.
+    /// Exactly one of them has to be non-NULL.
+    umf_memory_provider_handle_t upstream_memory_provider;
+
+    /// Init buffer used to pre-allocate memory at the creation time.
+    /// It has to be NULL if upstream_memory_provider is set.
+    /// Exactly one of them has to be non-NULL.
+    void *init_buffer;
+
+    /// Size of the pre-allocated buffer. If the `init_buffer` is set,
+    /// the `init_buffer_size` should be the size of this buffer.
+    size_t init_buffer_size;
+
+    /// When it is true and the upstream_memory_provider is given,
+    /// the init buffer (of `init_buffer_size` bytes) would be pre-allocated
+    /// during creation time using the `upstream_memory_provider`.
+    /// If upstream_memory_provider is not given,
+    /// the init_buffer is always used instead
+    /// (regardless of the value of this parameter).
+    bool immediate_init_from_upstream;
+
+    /// Set to true if umfMemoryProviderFree(upstream_memory_provider)
+    /// should never be called (for example if the upstream memory provider
+    /// does not support the free() operation).
+    bool disable_upstream_provider_free;
```

Comment:
> This might be something to fix in separate PR, but this API is bad experience. User must know properties of the memory provider, to correctly setup this provider.
> 

Yes, exactly - a user has to know the properties of the upstream memory provider. That's true.

> I think we should set free function ptr in the provider to NULL if it is not supported, and then we can detect in UMF if given provider support free or not and handle it internally

We should NOT do that automatically. A user must be aware of that - it must be their own decision.

Reply:
Done

Reply:
OK, I removed the input parameter and changed it to be detected automatically.